tapeout.1
.TH Tapeout 1 "17 March 1998" "Version 0.94"
.SH NAME
tapeout \- Output tapefiles to real tape
.SH SYNOPSIS
.B tapeout
.I sample_rate
.I tapefile
.RI [ start_block 
.RI [ output_file ]]
.SH DESCRIPTION
This program lets you copy a tapefile (
.I .tap 
or 
.I .tzx
) to real tape 
via soundcard. 
.PP
You have to connect the soundcard's 
.B line out
socket with the tape recorder's
.B mic
socket, then start this program. 
.SH OPERANDS
.TP
.B sample_rate
Sampling rate for the output. The higher this is, the better quality the
the result will be. It should be at least 16000.
.TP
.B tapefile
Name of the tape file to interpret, and output to the soundcard. It should
have either 
.I .tap
or 
.I .tzx
extension.
.TP
.B start_block
Specifies where to start playing the tapefile. Default is 0.
.TP
.B output_file
If given, the program writes the sound output into this file instead of the
soundcard. 
.SH SEE ALSO
xspect (1), vgaspect (1)
.PP
http://www.inf.bme.hu/~mszeredi/spectemu/spectemu.html
.SH BUGS
Many parameters are not changeable (like pause between tape blocks).
.PP
Not interactive (except for Ctrl\-c).
.SH AUTHOR
Miklos Szeredi (EMAIL: mszeredi@inf.bme.hu)
.PP
Tapeout is copyright (C) 1998 Miklos Szeredi
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version. See the file COPYING. 
.PP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.



ChangeLog
99/01/13  Version 0.95.3 released
=================================

99/01/12  Merged changes from Matan Ziv-Av:
	  - Joystick types
	  - 320x240x256V mode is used if available in vgaspect

99/01/11  Some bugs fixed

99/01/06  Merged changes from Matan Ziv-Av <zivav@cs.bgu.ac.il>:
	  - Snapshot and tapefile path
	  - Linux joystick support
	
99/01/05  Added PC-speaker support  

99/01/05  Virtual file support (with AVFS)

98/05/18  Version 0.94 released
===============================

98/05/18  Updated the man page (a more complete man-page is still needed!)
98/05/15  Some bugs fixed

98/05/14 Version 0.93.4 released
--------------------------------

98/05/14  Small additions:
	  - Keyboard matrix is emulated (only 1 level, I hope this is enough) 
          - "color refreshing" (maybe some other aplications could do it!)
          - Keyboard configuration is now fully implemented
	
98/05/11  Added DOS timer support with Allegro library

98/05/08  Version 0.93.3 released
---------------------------------
	
98/05/08  Created README.Z80 to describe how to use Z80 emulation	
98/05/06  Small new features + bugfixes
	  - Function keys for commonly used functions
	  - Keyboard config: (extended, spectrum, compat) and (shifted, raw) 
	    are now working. No custom keyboard or joystick yet
	  - Small modifications to make z80 emulation separate

98/04/30  Version 0.93.2 released
---------------------------------

98/04/30  I've removed the BROKEN_VGAKEYBOARD part from vgakey.c. 
	  I hope nobody's using SVGALIB version smaller than 1.2.11.

98/04/29  Some small new features: (see example.cfg for a complete list)
	  - different icon when the emulator is paused
	  - optional private colormap
	  - configurable colours (grayscale looks better on
	    monochrome displays)
	  - filetype can be given on the command line (tapefiles too)
	  - optional immediate loading of tapes (-load-immed)
	
98/04/27  Added configurability to spectemu:
	  - global config file (($prefix)/share/spectemu/spectemu.cfg)
	  - user's config file (~/.spectemurc)
	  - X resource database
	  - command line
	
98/04/23  Mode 320x240 is used in vgaspect (if available)

98/04/22  Version 0.93.1 released
---------------------------------
	
98/04/21  Handle "extra" ascii keys: []{}|\~
98/04/19  Added Spectrum keyboard picture, which shows pressed keys, and
	  keys can be pressed with mouse.
98/04/16  Rewrite keyboard handling:
	  - Fixes keyboard bug in X (key remained pressed)
	  - Better configurability: spectrum key definitions are in spkey.c
	    instead of xkey.c and vgakey.c

98/04/07  Added readline support by Christian Sievers <sievers@cip.rz.tu-bs.de>

98/03/24  Version 0.92 released
===============================
	
98/03/23  Last minute bugfixes
	
98/03/20  Version 0.92p3 released
98/03/20  Eliminated all sorts of minor bugs
98/03/19  Implement LOOP/CALL/RETURN blocks in TZX handling
98/03/17  Fixed (hopefully all) bugs in TZX handling:
 	  - There must be a 1ms pause at end of tape.
	  - Pausing after data blocks fixed.
	  - Made tape I/O timing ultra precise (e.g. see Flashpoint)
	
98/03/14  Fixed small bug in quick load
98/03/13  Sound of tape is now fixed (not perfect, but good enough)
          Created simple man page
	
98/03/11  Version 0.92pre2 released
98/03/10  Fixed quick load of tapefiles
98/03/10  Fixed resize bugs

98/03/09  Version 0.92pre1 released
98/03/08  C-version now handles bits 3 and 5 of the Flag register
	  (I hope I didn't break anything with this).
98/03/08  Wrote 'tapeout' utility to save tapefiles to real tape (via 
	  soundcard).

98/03/06  Implemented emulator pause.
98/03/06  Added ability to display spectrum keyboard layout on text 
	  console (Radovan.Garabik@fmph.uniba.sk).

98/03/05  Rewrite of snapshot handeling:
          - All versions of Z80 snapshots can now be loaded.
	  - Loading and saving of SNA snapshot format is supported. 
	  - Quick loading and saving of temporary snapshot.
	
98/03/03  Rewrite of tapefile handeling:
	  - Support for most TZX blocks (recognise length of all v1.11
	    blocks)
	  - SPT tapefiles are no longer supported 
          - Added quick loading of tapefiles
	  - Timing in tape emulation is very exact now, but this makes 
	    tape sound ugly (this will be fixed).
	
98/01/28  Version 0.91a released
98/01/26  Added detection for underlines in asm names to configure  

98/01/12  Version 0.91 released
===============================
	
98/01/08  Changes to configure to detect X libraries better
	
97/11/29  Use XSetWMProperties instead of XmbSetWMProperties
97/11/26  Created icon for xspect
97/11/20  IO and memory timing are improved (ULA / CPU clash)

97/11/18  Fixed 0xfe port problems
97/11/17  Input from unused ports (others than 0xFE) result in 0xFF, not 0x00

97/11/12  8000 sample/sec sound on SUN
97/11/12  Fixed screen bugs

97/11/11  Window resizing in X (thanks to Egmont)  
97/11/11  Fixed FocusIn/FocusOut bug (twm, olwm, ...)
97/11/11  Small changes and bugfixes
	
97/11/09  Sound on SUN (sample rate 16000, but 8000 is coming for older
	  SUNs.
	
97/10/26  'Interrupted system call' is not an error on sound device
97/10/26  Fixed screen scramble bug in console background running
	
97/10/20  Version 0.90pre1 released

97/10/20  Tested on the following Machines/Operating systems/Compilers:
	   o Intel/Linux/gcc              (assembly)
	   o Sparc/Solaris/gcc            (C)
	   o Alpha/OSF1/cc                (C)
	   o Sequent(i486)/Dynix-Ptx/gcc  (assembly)

97/10/20  Automatic configuration with 'configure' script   
97/10/20  Bugfixes and cleanups
97/10/20  Experimental DOS support with DJGPP (timer missing)
97/10/20  Added emulation in C
97/05/22  Background running for svgalib added (Needs svgalib >= 1.2.11)

97/04/17  Version 0.84 released
===============================
	
97/04/17  Added SLL to the emulated instructions

97/04/12  The following changes were inspired by Egmont Koblinger:
	  - Resize bug fixed
	  - Implemented tape file pause

tapeout.c
/* 
 * Copyright (C) 1998 Szeredi Miklos
 * Email: mszeredi@inf.bme.hu
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version. See the file COPYING. 
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <fcntl.h>
#include <unistd.h>

#ifdef HAVE_CONFIG_H
#  include "config.h"
#  if !defined(HAVE_SOUND) || !defined(OSS_SOUND)
#    define NO_SOUNDCARD
#  endif
#endif

#ifndef NO_SOUNDCARD
#include <sys/soundcard.h>
#include <sys/ioctl.h>
#define SOUND_DEV_NAME "/dev/dsp"
#else
#define SOUND_DEV_NAME "tape.out"
#endif

#include "tapefile.h"
#include "misc.h"

#define SBUFSIZE 1024
#define IMPBUFLEN 1024
#define LOWLEV 64
#define HIGLEV 192
#define LEVDIFF (HIGLEV - LOWLEV)

typedef unsigned short dbyte;
typedef unsigned char byte;

static void record_tapefile(int snd, int plen)
{
    static char soundbuf[SBUFSIZE];
    static dbyte impbuf[IMPBUFLEN];
    int i;
    long impsum;
    int impbufrem;
    int currlev;
    int osum;
    int segtype;
    int segctr;
    dbyte impl;
    dbyte *ibp = 0;
  

    impsum = 0;
    impbufrem = 0;
    currlev = 1;
    osum = 0;
  
    do {
        for(i = 0; i < SBUFSIZE; i++) {
            while(impsum < plen) {
                while(!impbufrem) {
                    impbufrem = next_imps(impbuf, IMPBUFLEN, plen * (SBUFSIZE - i));
                    if(!impbufrem) {
                        segtype = next_segment();
                        segctr = segment_pos();
                        if(segtype != SEG_VIRTUAL) printf("%4i: %s\n", segctr, seg_desc);
                        if(segtype <= SEG_END) {
                            write(snd, soundbuf, (unsigned) i);
                            return;
                        }
                    }
                    else ibp = impbuf;
                }
                currlev = !currlev;
        
                impl = *ibp;
                if(currlev) osum += impl;
                impsum += impl;
                ibp++;
                impbufrem--;
            }
            impsum -= plen;
            if(currlev) osum -= impsum;
            soundbuf[i] = (osum * LEVDIFF) / plen + LOWLEV;
            if(currlev) osum = impsum;
            else osum = 0;
        }
        write(snd, soundbuf, SBUFSIZE);
    } while(1);
}


int main(int argc, char *argv[])
{
    int snd;
    int res;
    int sound_sample_rate;
    int start_block;
    const char *sound_dev_name = SOUND_DEV_NAME;
    char *tapefile;
    int type;
    struct tape_options tapeopt;

    if(argc < 3) {
        fprintf(stderr, 
                "usage: %s sample_rate tapefile [start_block [output_file]]\n",
                argv[0]);
        exit(1);
    }
  
    sound_sample_rate = atoi(argv[1]);
    tapefile = argv[2];
  
    start_block = 0;
    if(argc > 3) start_block = atoi(argv[3]);
    if(argc > 4) sound_dev_name = argv[4];

    snd = open(sound_dev_name, O_WRONLY | O_CREAT, 0644);

    if(snd < 0) {
        fprintf(stderr, "Could not open sound device '%s': %s\n", 
                sound_dev_name, strerror(errno));
        exit(1);
    }

#ifndef NO_SOUNDCARD
    {
        int parm;

        parm = sound_sample_rate;
        res = ioctl(snd,SOUND_PCM_WRITE_RATE,&parm);
        if(res < 0) 
            fprintf(stderr, "ioctl(SOUND_PCM_WRITE_RATE, %i) failed: %s\n", 
                    sound_sample_rate, strerror(errno));
        sound_sample_rate = parm;
    
        res = ioctl(snd,SOUND_PCM_SYNC,0);
        if(res < 0) 
            fprintf(stderr, "ioctl(SOUND_PCM_WRITE_SYNC, 0) failed: %s\n", 
                    strerror(errno));
    }
#endif

    if(check_ext(tapefile, "tap")) type = TAP_TAP;
    else if(check_ext(tapefile, "tzx")) type = TAP_TZX;
    else {
        fprintf(stderr, "tapefile must have either '.tap' or '.tzx' extension\n");
        exit(1);
    }

    res = open_tapefile(tapefile, type);
    if(!res) {
        fprintf(stderr, "%s\n", seg_desc);
        exit(1);
    }
    INITTAPEOPT(tapeopt);
    tapeopt.blanknoise = 0;
    tapeopt.stoppause = 5;
    set_tapefile_options(&tapeopt);

    if(start_block) {
        res = goto_segment(start_block);
        if(!res) {
            fprintf(stderr, "Could not jump to block %i: %s\n", start_block, 
                    seg_desc);
            exit(1);
        }
    }

    fprintf(stderr, "Recording `%s' to audio output at %i sample rate\n", 
            tapefile, sound_sample_rate);

    record_tapefile(snd, 3500000 / sound_sample_rate);

    return 0;
}
